utils/auth.ts

```diff
@@ -1,33 +1,98 @@
+import { Lucia } from 'lucia';
 import { prisma as client } from '~/utils/db';
-import 'lucia/polyfill/node'; // polyfill for Node.js versions <= 18
-import * as context from 'next/headers';
-import { nextjs_future } from 'lucia/middleware';
+import { PrismaAdapter } from '@lucia-auth/adapter-prisma';
 import { env } from '~/env.mjs';
-import type { User } from '@prisma/client';
 import { cache } from 'react';
+import { redirect } from 'next/navigation';
+import { cookies } from 'next/headers';
+import 'server-only';
 
-export const auth = lucia({
-  env: env.NODE_ENV === 'production' ? 'PROD' : 'DEV',
-  middleware: nextjs_future(),
+const adapter = new PrismaAdapter(client.session, client.user);
+
+export const lucia = new Lucia(adapter, {
   sessionCookie: {
+    name: 'fresco-session',
+    // this sets cookies with super long expiration
+    // since Next.js doesn't allow Lucia to extend cookie expiration when rendering pages
     expires: false,
+    attributes: {
+      // set to `true` when using HTTPS
+      secure: env.NODE_ENV === 'production',
+    },
   },
-  getUserAttributes: (data: User) => {
+});
+
+declare module 'lucia' {
+  interface Register {
+    Lucia: typeof lucia;
+    // DatabaseSessionAttributes: DatabaseSessionAttributes;
+    DatabaseUserAttributes: DatabaseUserAttributes;
+  }
+}
+
+// interface DatabaseSessionAttributes {}
+interface DatabaseUserAttributes {
+  username: string;
+  hashedPassword: string;
+}
+
+export const getServerSession = cache(async () => {
+  const sessionId = cookies().get(lucia.sessionCookieName)?.value ?? null;
+  if (!sessionId)
     return {
-      username: data.username,
+      session: null,
+      user: null,
     };
-  },
-  adapter: prismaAdapter(client),
-  // experimental: {
-  //   debugMode: env.NODE_ENV !== 'production',
-  // },
+  const result = await lucia.validateSession(sessionId);
+  try {
+    if (result.session && result.session.fresh) {
+      const sessionCookie = lucia.createSessionCookie(result.session.id);
+      cookies().set(
+        sessionCookie.name,
+        sessionCookie.value,
+        sessionCookie.attributes,
+      );
+    }
+    if (!result.session) {
+      const sessionCookie = lucia.createBlankSessionCookie();
+      cookies().set(
+        sessionCookie.name,
+        sessionCookie.value,
+        sessionCookie.attributes,
+      );
+    }
+  } catch {
+    // Next.js throws error when attempting to set cookies when rendering page
+  }
+  return result;
 });
 
-export type Auth = typeof auth;
+export type RequireAuthOptions = {
+  redirectPath?: string;
+};
 
-export const getServerSession = cache(() => {
-  const authRequest = auth.handleRequest('GET', context);
-  return authRequest.validate();
-});
+export async function requirePageAuth(
+  { redirectPath = null } = {} as RequireAuthOptions,
+) {
+  const { session } = await getServerSession();
+
+  if (!session) {
+    if (!redirectPath) {
+      redirect('/signin');
```

Comment:
why is this redirect being called in an if statement when it's going to be called regardless of the condition right after?